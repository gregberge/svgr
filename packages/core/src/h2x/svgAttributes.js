import { JSXAttribute } from 'h2x-plugin-jsx'

const compareAttrsName = attr => otherAttr => attr.name === otherAttr.name
const compareAttrsValue = attr => otherAttr => attr.value === otherAttr.value

const compareAttrs = attr => otherAttr =>
  compareAttrsName(attr)(otherAttr) && compareAttrsValue(attr)(otherAttr)

const areAttrsAlreadyInjected = (node, attributes = {}) => {
  const nodeAttrs = node.attributes

  return Object.keys(attributes).reduce((accumulation, key) => {
    if (nodeAttrs.some(compareAttrs({ name: key, value: attributes[key] })))
      return accumulation
    return false
  }, true)
}

const svgAttribute = (attributes = {}) => () => {
  const keys = Object.keys(attributes)

  return {
    visitor: {
      JSXElement: {
        enter(path) {
          if (path.node.name !== 'svg') return
          if (areAttrsAlreadyInjected(path.node, attributes)) return

          const parseAttributes = keys.reduce((accumulation, key) => {
            const props = new JSXAttribute()
            props.name = key
            props.value = attributes[key]
            return [...accumulation, props]
          }, [])

          const mergeAttributes = path.node.attributes.reduce(
            (accumulation, value) => {
              if (accumulation.some(compareAttrsName(value)))
                return accumulation
              return [...accumulation, value]
            },
            parseAttributes,
          )

          path.node.attributes = mergeAttributes
          path.replace(path.node)
        },
      },
    },
  }
}

export default svgAttribute
