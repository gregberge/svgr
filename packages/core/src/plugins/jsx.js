import unified from 'unified'
import parse from 'rehype-parse'
import vfile from 'vfile'
import hastToBabelAst from 'hast-util-to-babel-ast'
import { transformFromAstSync, createConfigItem } from '@babel/core'
import removeJSXAttribute from '../babelPlugins/removeJSXAttribute'
import removeJSXEmptyExpression from '../babelPlugins/removeJSXEmptyExpression'

export default (code, config, state = {}) => {
  const filePath = state.filePath || 'unknown'
  const hastTree = unified()
    .use(parse, {
      fragment: true,
      space: 'svg',
      emitParseErrors: true,
      duplicateAttribute: false,
    })
    .parse(vfile({ path: filePath, contents: code }))

  const babelTree = hastToBabelAst(hastTree)

  const { code: transformedCode } = transformFromAstSync(babelTree, code, {
    plugins: [
      createConfigItem([
        removeJSXAttribute,
        { attribute: 'xmlns' },
        'removeXmlns',
      ]),
      createConfigItem(removeJSXEmptyExpression),
    ],
    filename: filePath,
    babelrc: false,
    configFile: false,
  })

  return transformedCode
}
