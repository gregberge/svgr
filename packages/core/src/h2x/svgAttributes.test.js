import jsx from 'h2x-plugin-jsx'
import { transform } from 'h2x-core'
import svgAttributes from './svgAttributes'

describe('svgAttributes', () => {
  it('should add one attribute to svg', () => {
    const attrs = { focusable: false }
    const result = transform(
      `
      <svg>
        <g stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
        </g>
      </svg>
    `,
      { plugins: [jsx, svgAttributes(attrs)] },
    )

    expect(result).toMatchSnapshot()
  })

  it('should add multiple attributes to svg', () => {
    const attrs = {
      focusable: false,
      hidden: 'hidden',
    }
    const result = transform(
      `<svg>
        <g stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
        </g>
      </svg>
    `,
      { plugins: [jsx, svgAttributes(attrs)] },
    )

    expect(result).toMatchSnapshot()
  })

  it('should update old attribute to the new attribute', () => {
    const attrs = { focusable: true }
    const result = transform(
      `<svg focusable="false">
        <g stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
        </g>
      </svg>
    `,
      { plugins: [jsx, svgAttributes(attrs)] },
    )

    expect(result).toMatchSnapshot()
  })

  it('should add a new attribute', () => {
    const attrs = {
      focusable: true,
      hidden: true,
    }
    const result = transform(
      `<svg focusable="true" color="#fff">
        <g stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
        </g>
      </svg>
    `,
      { plugins: [jsx, svgAttributes(attrs)] },
    )

    expect(result).toMatchSnapshot()
  })
})
