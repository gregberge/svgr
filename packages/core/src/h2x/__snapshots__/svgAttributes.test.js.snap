// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svgAttributes should add a new attribute 1`] = `
"<svg focusable=\\"true\\" hidden=\\"true\\" color=\\"#fff\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;

exports[`svgAttributes should add multiple attributes to svg 1`] = `
"<svg focusable=\\"false\\" hidden=\\"hidden\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;

exports[`svgAttributes should add one attribute to svg 1`] = `
"<svg focusable=\\"false\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;

exports[`svgAttributes should update old attribute to the new attribute 1`] = `
"<svg focusable=\\"true\\">
  <g stroke=\\"#063855\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;
